test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs

```diff
@@ -319,6 +353,108 @@ public void BulkUpdateTwoTopLevelObjects_WithTheSameNestedObject_DeserializesSuc
             Assert.Equal(carTargetEntity, car2TargetEntity);
         }
 
+        [Fact]
+        public async Task BulkUpdateAsyncTwoTopLevelObjects_WithTheSameNestedObject_DeserializesSuccessfully()
+        {
+            var expectedResponse = "{\"@context\":\"http://localhost:8000/$metadata#Persons/$delta\",\"value\":[{\"ID\":100,\"Name\":\"Bing\",\"Cars@delta\":[{\"@id\":\"http://localhost:8000/Cars(1001)\"}]},{\"ID\":200,\"Name\":\"Edge\",\"Cars@delta\":[{\"@id\":\"http://localhost:8000/Cars(1001)\"}]}]}";
+
+            SetupContextWithRequestPipelineForSaving(
+                this.context,
+                expectedResponse);
+
+            var person = new Person
+            {
+                ID = 100,
+                Name = "Bing",
+            };
+
+            var person2 = new Person
+            {
+                ID = 200,
+                Name = "Edge"
+            };
+
+            var car = new Car
+            {
+                ID = 1002,
+                Name = "CarA"
+            };
+
+            this.context.AttachTo("Persons", person);
+            this.context.AttachTo("Persons", person2);
+            this.context.AttachTo("Cars", car);
+
+            this.context.AddLink(person, "Cars", car);
+            this.context.AddLink(person2, "Cars", car);
+
+            DataServiceResponse response = await this.context.BulkUpdateAsync(person, person2);
+
+            var personOperationResponse = response.First() as ChangeOperationResponse;
+            var person1 = (personOperationResponse.Descriptor as EntityDescriptor).Entity as Person;
+            var personcarOperationResponse = personOperationResponse.NestedResponses.FirstOrDefault() as ChangeOperationResponse;
+            var carTargetEntity = (personcarOperationResponse.Descriptor as LinkDescriptor).Target as Car;
+            var person2OperationResponse = response.Last() as ChangeOperationResponse;
+            var person2Response = (person2OperationResponse.Descriptor as EntityDescriptor).Entity as Person;
+            var person2carOperationResponse = personOperationResponse.NestedResponses.FirstOrDefault() as ChangeOperationResponse;
+            var car2TargetEntity = (person2carOperationResponse.Descriptor as LinkDescriptor).Target as Car;
+
+            Assert.Equal(2, response.Count());
+            Assert.Equal("Bing", person1.Name);
+            Assert.Equal("Edge", person2Response.Name);
+            Assert.Equal(carTargetEntity, car2TargetEntity);
+        }
+
+        [Fact]
+        public async Task NullTopLevelObjects_ShouldThrowException()

```

Comment:
This test name seems inconsistent with what it's actually doing. It doesn't seem to be testing top-level nulls.
Also, in addition to testing for null top-level objects, would also be great to test exceptions that occur inside the async request processing to ensure the async exception handling logic introduced in this PR works as expected.

Reply:
+1

Reply:
Here is an example of a test: https://github.com/OData/odata.net/blob/52487c2c2c620156d53c75909828327133be8210/test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/BulkUpdateE2ETests.cs#L203